### PR TITLE
Add style exclusions. Closes #113

### LINKED ISSUE #113: User-friendly exclusions
How to exclude certain sites from global styles is still one of the most FAQs on a regular basis. Some users will just never get a basic grasp of regex, even if explained in depth. I know I've seen a PR from back in the day where you argued for this pretty adamantly @tophf , so I'm kinda surprised you never brought it up here. 

Jason's main argument was the differences in Chrome/FF versions. Not that I agree with that argument, but Stylus is the same in either regardless. IIRC, you had one that was at least semi-implemented to be integrated in editor pages. I don't know how difficult it'd be, but it'd be much more user-friendly to have it in the popup if that's not unrealistic. An exclude/undo button next to globals would be ideal if possible.

### PR DESCRIPTION
**Pull request by Mottie**
* Popup: Right click on a style's check box to open exclusions menu for a page.
* Manage page: "Excluded on: ..." added to "Applies to" column (in both old and new UI).
* Edit page: Excluded Pages section added to side bar.

### CODE REVIEW

**Comment by JourneyOver** [comment]
So far this looks pretty nice :D

one bug I've run into so far though is that if I an exclusion through the edit style page and then go and add one through the popup menu it goes and deletes any excluded sites I had added through the edit style page.

**Comment by Mottie** [comment]
The popup menu select should highlight an option if it's already in the exclusion list. If you remove it in the popup, it also removes it from the exclusion list.

For example, if you right click the checkbox on google.com, you should see select options with "Domain: google.com", "Domain: www.google.com" and "Prefix: https://www.google.com". If you already have "google.com" in the list, it should be highlighted. If you remove it from the select, it'll remove it from the exclusion list. You can also select multiple options; I know this will add duplicates, but you can change them from the edit page for more fine-tuning.

**Comment by Mottie** [comment]
Solved in chat: Make sure to save any exclusions you add. They don't automatically save with the style.

**Comment by JourneyOver** [comment]
Will remember that for now on haha :P

**Comment by tophf** [comment]
Popup:

* The modal in the popup is too big, it also has big margins. I think it should be similar to the confirmation on deleting a style. No need to show the style name, you can highlight it somehow in the list e.g. by adding a class like `selected` or `highlighted` which has a border or an underline.
* `Edit excluded page(s)` is misleading: I would expect being able to actually edit the value. `Exclude the site or page` maybe.

Editor:

* `Excluded Pages: N sites` title contradicts itself: sites or pages? I think we can strip the extra word by using an additional i18n message with a placeholder: `Excluded $num$ pages`. The number won't be highlighted.
* When there are exclusions the section should probably always expand on opening the editor, conversely when there are none it should always be collapsed on opening the editor.
* `No exclusions` shouldn't have a border because it's not an element. Arguably, there's no need to display any text at all.
* `Add excluded page` should just add a new input box in the list and position the caret there, no popups.
* An `x` icon to the right of each entry to delete it would be more convenient and obvious.
* Editing should be possible via click on an entry. Entries should be `input` elements for that.
* Not sure why regexp isn't supported, I think it should be, for example using the `/re/` notation.
* I can't exclude `*://*.google.co*/*foo*` - the asterisk should be handled in any place.
* In addition to the info icon with its lengthy explanation it'd be nice to have a one line example right next to the now single `Add` button e.g. `*://*.google.co*/*foo*`

**Comment by JourneyOver** [comment]
This still being worked on? Would love to see this get implemented into the stylus fully..

**Comment by Mottie** [comment]
Yeah, I am... on and off (prepping a baby room 👶). I ended up almost totally rewriting it and I still have one more issue to work out before I update this PR.

**Comment by JourneyOver** [comment]
Ahh sounds fun :p and glad this is still being worked on. Can't wait for this to get into the actual releases of stylus as so far I've just been running the latest releases while adding in the exclusion branch stuff myself to stay up to date but having to manually do it get's tedious after a while.

**Comment by Mottie** [comment]
Still not working yet... I wanted to push the current changes here because my hard drive has been acting flaky.

Last issue I need to address (that I remember) is getting the popup to update all tabs after adding/removing exclusions; the database gets updated, just not the active tabs.

**Comment by JourneyOver** [comment]
Got it! thanks for letting me know. :)

**Comment by Mottie** [comment]
Thanks @JourneyOver, I've fixed the expanding popup; but I've been having a lot of difficulty with getting the exclusions to update dynamically.

Previously, when it sorta worked, excluding a style would completely remove it from the list of styles in the popup. Now, the excluded style is visible & red; but updating it for all tabs is just not working. I think I'm going to need @tophf to look at the code and provide some suggestions.

**Comment by JourneyOver** [comment]
@Mottie it looks like whatever you did recently fixed the problem I had noted above about it not adding the exclusion when using the exclusion popup (it adds the exclusion to the exclusion list for the theme along with marking the theme red in the popup for said tabs with the excluded site now)

**Comment by JourneyOver** [comment]
Please tell me this hasn't been forgotten about and is still being worked on.. So waiting for this to be included in stylus by default :s

**Comment by Mottie** [comment]
It's not forgotten, but it is on the back-burner... I haven't had much time to mess with the branch.

**Comment by Mottie** [comment]
@JourneyOver Please test the latest updates.

**Comment by JourneyOver** [comment]
Glad to see new updates to this :D

---------

Bugs:

1. After adding a site exclusion to a theme it seems to make toolbar icon stop showing the number of active themes for said site. Example -- https://i.imgur.com/4TIzOXq.png I have 1 active theme for the site still, but the icon is showing as if there are none. Refreshing does not seem fix it. 

Going in removing exclusion and re-adding it seems to somewhat fix the toolbar icon, it stills end up being greyed out but it shows the active styles number as well. Example -- https://i.imgur.com/VoHismF.png refreshing the page brings back the color to the icon completely.

2. Going into a style and adding an exclusion manually instead of going to site and doing it through the dropdown, while adding the site through the excluded it will randomly not stop focusing in the text box area so you'll have to refocus multiple times if the url is going to be long.

3. Sometimes a style with a site excluded will just randomly appear in the dropdown menu and sometimes it will disappear..

Example -- 
Install style https://userstyles.org/styles/142759/darkerr-a-darker-theme-for-sonarr-radarr 
Add exclusion "trakt.tv" to style
Go to https://trakt.tv/home (theme will appear in the dropdown)
Go to https://trakt.tv/dashboard (theme will not appear in the dropdown)
(you'll need an account to view dashboard (but I'm sure there are other sites out there that could end up showing the same bug that doesn't require a sign up))

4. Adding an exclusion and then removing said exclusion doesn't re-apply the theme to the site, refreshing the tab sometimes does not fix it, while sometimes it does fix it.


Those are at least a few of the bugs I've noticed in the latest update. If I run into anymore I will post them up.

**Comment by JourneyOver** [comment]
Seen you updated some stuff so testing latest version after fixing the conflicting file (which seems to be just a one line fix).. I'll be quoting a few of the issues above that are fixed/not fixed along with reporting any new things that have cropped up..

> After adding a site exclusion to a theme it seems to make toolbar icon stop showing the number of active themes for said site. Example -- i.imgur.com/4TIzOXq.png I have 1 active theme for the site still, but the icon is showing as if there are none. Refreshing does not seem fix it.

Seems to still be broken, and even more broken than before it seems actually.. I still get the issue that I have in the picture, along with it now just randomly deciding to stop showing active theme count when visiting a site (without doing the whole exclusion thing) (refreshing doesn't seem to fix this every time but sometimes it will) 

At times it'll show the wrong theme count as well, for example I have 3 themes active but it'll say 2 or 1 instead of 3. This one doesn't happen a lot but I see it every so often.

Another ontop of this seems that the icon color seems to almost vanish as well https://i.imgur.com/0ftZfdq.png (My default color icon supposed to look like https://i.imgur.com/VSRzPCP.png), again most of the time a refresh of the tab seems to mostly fix this.

> Going into a style and adding an exclusion manually instead of going to site and doing it through the dropdown, while adding the site through the excluded it will randomly stop focusing in the text box area so you'll have to refocus multiple times if the url is going to be long.

This seems to be fixed, as I am able to type in the exclusion text box forever now without it unfocusing

> Sometimes a style with a site excluded will just randomly appear in the dropdown menu and sometimes it will disappear..
> 
> Example --
> Install style userstyles.org/styles/142759/darkerr-a-darker-theme-for-sonarr-radarr
> Add exclusion "trakt.tv" to style
> Go to trakt.tv/home (theme will appear in the dropdown)
> Go to trakt.tv/dashboard (theme will not appear in the dropdown)
> (you'll need an account to view dashboard (but I'm sure there are other sites out there that could end up showing the same bug that doesn't require a sign up))
> 
> before it never appeared in the dropdown unless there was a sonarr or radarr in the url somewhere, but now for the most part is just shows up in the dropdown through the whole site for the most part if the site is in the excluded.

This still seems to be an issue, in that it'll now show up for the whole site unlike what I had explained in the example bit above.. If this is the intended way you plan on doing it, I'm still thinking it should go back to how it used to be tbh where the excluded theme only showed up in the dropdown if the url contained something from the applies to url of the theme..

> Adding an exclusion and then removing said exclusion doesn't re-apply the theme to the site, refreshing the tab sometimes does not fix it, while sometimes it does fix it.

Not fixed, though it does seem like I can now tick/untick the theme to make it re-apply unlike before where it wouldn't and it'd take some refreshing of the tab to fix it.

-----

I haven't really run into anything else really new yet, but I'm keeping an eye out.

File: background/storage.js [hunk 1]
```diff
@@ -309,9 +311,11 @@ function filterStylesInternal({
           stopOnFirst: !asHash,
           skipUrlCheck: true,
           matchUrlBase,
+          omitCode,
         });
       if (asHash) {
-        if (sections.length) {
+        if (sections.length && sections[0] !== '') {
+          sections.unshift({included: !isPageExcluded(matchUrl, style.exclusions)});
```
[the remark this answers is not in the document]
Comment (Mottie) [reply background/storage.js]: Yeah I know :(... it was a workaround to get the style name to show in the popup when excluded so the list of exclusions could be modified.

File: popup/popup-exclusions.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    });
+    // Remove exclusions (unselected in popup modal)
+    all.forEach(value => {
+      exclusionExists(current, value).forEach(ending => {
+        delete style.exclusions[value + ending];
+      });
+    });
+    exclusions.save({
+      id: style.id,
+      exclusionList: style.exclusions
```
[the remark this answers is not in the document]
Comment (Mottie) [reply popup/popup-exclusions.js]: I think it all gets JSON stringified anyway.

Honestly, feel free to close this PR and completely rewrite it. I think your suggestions to simplify the API are great! It's currently very complicated & confusing.

**Comment by JourneyOver** [comment]
Glad to see this is getting worked on again, currently don't have a computer anymore but I'll start testing things again once I do (hopefully by the end of the week)

**Comment by eight04** [comment]
It is still WIP. I'm rewriting the entire backend. It won't work until the message system is completely moved to the new system. Also, part of the system requires #510.

File: background/background.js [hunk 3]
```diff
@@ -295,99 +297,34 @@ window.addEventListener('storageReady', function _() {
...
-      getTab(sender.tab.id),
-    ]).then(([styles, tab]) => {
-      if (tab) styles.exposeIframes = tab.url.replace(/(\/\/[^/]*).*/, '$1');
-      return styles;
-    });
-  };
-  const updateAPI = (_, enabled) => {
-    window.API_METHODS.getStylesForFrame = enabled ? getStylesForFrame : getStyles;
-  };
-  prefs.subscribe(['exposeIframes'], updateAPI);
```
Comment (eight04) [inline background/background.js]: This should be fixed in #510.

Comment (eight04) [reply background/background.js]: It needs a new API like `API.getTabDomain` to fetch the domain of the parent window. I'll fix this after #510 is merged.

**Comment by JourneyOver** [comment]
How are things going with this?

**Comment by JourneyOver** [comment]
@eight04 since #510 has been merged in from the looks of it, any idea when you might possibly get back to this? As there are now too many changes to master that I sadly can't go about just merging changes and continuing to use exclusions :/ so I am now stuck in the past on stylus versions as I kinda rely on exclusions for various styles atm.

**Comment by eight04** [comment]
We have to reimplement this on the current master (maybe after fixing all 1.5.x bugs). Also, @Mottie is designing a new style manager, so I guess we can first add the exclusion dialog to the popup and the editor.

**Comment by SimpleUs** [comment]
Sorry, but where can I download this version of the extension, where do exceptions work.

**Comment by eight04** [comment]
Anyone still looking forward to this may take a look at #666.

**Pull request closed without merging**